packages/@aws-cdk/custom-resources/lib/provider-framework/runtime/framework.ts

```diff
@@ -0,0 +1,177 @@
+// tslint:disable: no-console
+// tslint:disable: max-line-length
+import { IsCompleteResponse, OnEventResponse } from '../types';
+import { submitCloudFormationResponse } from './cfn-response';
+import consts = require('./consts');
+import { invokeFunction, startExecution } from './outbound';
+import { failOnError, getEnv, log, Retry } from './util';
+
+// use consts for handler names to compiler-enforce the coupling with construction code.
+export = {
+  [consts.FRAMEWORK_ON_EVENT_HANDLER_NAME]: onEvent,
+  [consts.FRAMEWORK_IS_COMPLETE_HANDLER_NAME]: isComplete,
+  [consts.FRAMEWORK_ON_TIMEOUT_HANDLER_NAME]: onTimeout
+};
+
+/**
+ * The main runtime entrypoint of the async custom resource lambda function.
+ *
+ * Any lifecycle event changes to the custom resources will invoke this handler, which will, in turn,
+ * interact with the user-defined `onEvent` and `isComplete` handlers.
+ *
+ * This function will always succeed. If an error occurs
+ *
+ * @param cfnRequest The cloudformation custom resource event.
+ */
+async function onEvent(cfnRequest: AWSLambda.CloudFormationCustomResourceEvent) {
+  await failOnError(cfnRequest, async () => {
+    log('onEventHandler', cfnRequest);
+
+    cfnRequest.ResourceProperties = cfnRequest.ResourceProperties || { };
+
+    // validate incoming request
+    validateCfnRequest(cfnRequest);
+
+    const onEventResult = await invokeUserFunction(consts.USER_ON_EVENT_FUNCTION_ARN_ENV, cfnRequest) as OnEventResponse;
+    log('onEvent returned:', onEventResult);
+
+    // merge the request and the result from onEvent to form the complete resource event
+    // this also performs validation.
+    const resourceEvent = createResponseEvent(cfnRequest, onEventResult);
```

Comment:
If this throws, we created something but we're not going to clean it up. Okay?

Reply:
I guess the state of the resource is not ensured if the operation fails in-air (either within onEvent, within the framework or within isComplete). Not sure there's anything smart we can do about this. What do you think?

Reply:
Maybe not, but please put a comment to that effect in there so in the future we'll at least know it was considered.

Reply:
added a README section